### PR TITLE
add objectness score to detection

### PR DESCRIPTION
**Pull request by 201power**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds an explicit objectness score to YOLO detection models, with training, inference, NMS, results, and plotting all updated to support it. 🚀

### 📊 Key Changes
- New objectness head and output
  - Detection head now predicts an additional obj score (sigmoid) per box.
  - Model output shape becomes [x, y, w, h, class_probs, obj].
  - Bias initialized for obj to a low prior probability.
- Training and loss
  - New loss component: obj_loss using BCE.
  - Loss names updated to include obj_loss.
  - New hyperparameter `obj` (default 0.0) controls obj loss weighting in `default.yaml`.
- Inference and NMS
  - `model.predict()` adds `obj` arg (default 0.0) to filter detections by objectness.
  - NMS now supports `obj_thres` and filters by both class conf and objectness.
  - Postprocess/top-k selection retains and returns obj scores.
- Results API changes
  - Results boxes now include objectness; shape changes from 6→7 (or 7→8 when tracking).
  - New `Boxes.obj` property; `Boxes.conf`, `Boxes.cls`, and `Boxes.id` indices adjusted accordingly.
- Visualization
  - Training plots expanded to include new metrics layout (e.g., obj_loss).
- Misc
  - Minor update in pretrained weight loading logic (channel transfer safety).

### 🎯 Purpose & Impact
- Better detection control: filter predictions using objectness for cleaner outputs and fewer false positives. 🎯
- Training flexibility: opt-in objectness learning (`hyp.obj > 0`) can improve stability and precision on certain datasets.
- Richer outputs: users can now access per-box objectness via `results.boxes.obj`.
- Logging and monitoring: obj_loss now visible in training metrics and plots.
- Note on compatibility:
  - If you parse raw predictions or rely on fixed result tensor sizes, update code for the extra obj column.
  - Boxes now expect 7 or 8 values (including obj). Older assumptions of 6/7 may need adjustment.

Quick usage tips:
- Filter by objectness at inference:
  ```python
  from ultralytics import YOLO
  model = YOLO('yolo11n.pt')
  results = model.predict('image.jpg', conf=0.25, obj=0.3)
  ```
- Access objectness scores:
  ```python
  boxes = results[0].boxes
  obj_scores = boxes.obj  # tensor of per-box objectness
  ```

See the Ultralytics Docs for details.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @201power, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! Your contribution to enhancing the YOLO model with an objectness head is exciting and much appreciated. To ensure a smooth integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features you’ve introduced.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your feature addition or bug fix. _"Perfection is achieved not when there is nothing more to add, but when there is nothing left to take away."_ — Antoine de Saint-Exupéry

For more detailed guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions, don’t hesitate to leave a comment, and an Ultralytics engineer will assist you shortly. 🚀🤖 

Thank you for helping make Ultralytics even better! 🌟
